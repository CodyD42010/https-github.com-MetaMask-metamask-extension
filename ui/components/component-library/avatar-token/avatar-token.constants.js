import { Size } from '../../../helpers/constants/design-system';

export const AVATAR_TOKEN_SIZES = {
  XS: Size.XS,
  SM: Size.SM,
  MD: Size.MD,
  LG: Size.LG,
  XL: Size.XL,
};
