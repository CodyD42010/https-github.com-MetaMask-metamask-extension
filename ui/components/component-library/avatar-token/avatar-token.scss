.avatar-initial {
  font-size: var(--typography-s-body-sm-bold-font-size);
  font-weight: var(--typography-s-body-sm-bold-font-weight);

  @include screen-sm-min {
    font-size: var(--typography-l-body-sm-bold-font-size);
  }
}
