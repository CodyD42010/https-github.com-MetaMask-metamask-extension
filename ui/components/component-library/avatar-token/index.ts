export { AvatarToken } from './avatar-token';
export type { AvatarTokenProps } from './avatar-token.types';
