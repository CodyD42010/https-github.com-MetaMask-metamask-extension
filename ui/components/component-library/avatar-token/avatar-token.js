import React, { useState, useEffect } from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';
import Box from '../../ui/box/box';
import { AvatarBase } from '../avatar-base';
import {
  Size,
  DISPLAY,
  AlignItems,
  JustifyContent,
  TextColor,
  BorderColor,
  BackgroundColor,
} from '../../../helpers/constants/design-system';
import { AVATAR_TOKEN_SIZES } from './avatar-token.constants';

export const AvatarToken = React.forwardRef(
  (
    {
      size = Size.MD,
      name,
      src,
      showHalo,
      color = TextColor.textDefault,
      backgroundColor = BackgroundColor.backgroundAlternative,
      borderColor = BorderColor.transparent,
      className,
      ...props
    },
    ref,
  ) => {
    const [showFallback, setShowFallback] = useState(false);

    useEffect(() => {
      setShowFallback(!src);
    }, [src]);

    const handleOnError = () => {
      setShowFallback(true);
    };

    const fallbackString = name && name[0] ? name[0] : '?';

    return (
      <AvatarBase
        ref={ref}
        size={size}
        display={DISPLAY.FLEX}
        alignItems={AlignItems.center}
        justifyContent={JustifyContent.center}
        className={classnames(
          'mm-avatar-token',
          showHalo && 'mm-avatar-token--with-halo',
          className,
        )}
        {...{ backgroundColor, borderColor, color, ...props }}
      >
        {showFallback ? (
          fallbackString
        ) : (
          <>
            {showHalo && (
              <img
                src={src}
                className={
                  showHalo ? 'mm-avatar-token__token-image--blurred' : ''
                }
                aria-hidden="true"
              />
            )}
            <img
              className={
                showHalo
                  ? 'mm-avatar-token__token-image--size-reduced'
                  : 'mm-avatar-token__token-image'
              }
              onError={handleOnError}
              src={src}
              alt={`${name} logo` || 'token logo'}
            />
          </>
        )}
      </AvatarBase>
    );
  },
);

AvatarToken.propTypes = {
  /**
   * The name accepts the string to render the first letter of the AvatarToken. This will be used as the fallback display if no image url is passed to the src
   */
  name: PropTypes.string,
  /**
   * The src accepts the string of the image to be rendered
   */
  src: PropTypes.string,
  /**
   * The showHalo accepts a boolean prop to render the image with halo effect
   */
  showHalo: PropTypes.bool,
  /**
   * The size of the AvatarToken.
   * Possible values could be SIZES.XS(16px), SIZES.SM(24px), SIZES.MD(32px), SIZES.LG(40px), SIZES.XL(48px)
   * Defaults to SIZES.MD
   */
  size: PropTypes.oneOf(Object.values(AVATAR_TOKEN_SIZES)),
  /**
   * The background color of the AvatarToken
   * Defaults to Color.backgroundAlternative
   */
  backgroundColor: PropTypes.oneOf(Object.values(BackgroundColor)),
  /**
   * The background color of the AvatarToken
   * Defaults to Color.borderDefault
   */
  borderColor: PropTypes.oneOf(Object.values(BorderColor)),
  /**
   * The color of the text inside the AvatarToken
   * Defaults to Color.textDefault
   */
  color: PropTypes.oneOf(Object.values(TextColor)),
  /**
   * Additional classNames to be added to the AvatarToken
   */
  className: PropTypes.string,
  /**
   * AvatarToken also accepts all Box props including but not limited to
   * className, as(change root element of HTML element) and margin props
   */
  ...Box.propTypes,
};

AvatarToken.displayName = 'AvatarToken';
