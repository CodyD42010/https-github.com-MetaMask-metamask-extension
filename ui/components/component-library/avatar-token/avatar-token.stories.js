import React from 'react';
import {
  Size,
  DISPLAY,
  AlignItems,
  TextColor,
  BackgroundColor,
  BorderColor,
} from '../../../helpers/constants/design-system';

import Box from '../../ui/box/box';

import {
  AvatarNetwork,
  BUTTON_LINK_SIZES,
  BadgeWrapper,
  ButtonLink,
  Text,
} from '..';

import README from './README.mdx';

import { AvatarToken } from './avatar-token';
import { AVATAR_TOKEN_SIZES } from './avatar-token.constants';

export default {
  title: 'Components/ComponentLibrary/AvatarToken',

  component: AvatarToken,
  parameters: {
    docs: {
      page: README,
    },
  },
  argTypes: {
    size: {
      control: 'select',
      options: Object.values(AVATAR_TOKEN_SIZES),
    },
    color: {
      options: Object.values(TextColor),
      control: 'select',
    },
    backgroundColor: {
      options: Object.values(BackgroundColor),
      control: 'select',
    },
    borderColor: {
      options: Object.values(BorderColor),
      control: 'select',
    },
    name: {
      control: 'text',
    },
    src: {
      control: 'text',
    },
    showHalo: {
      control: 'boolean',
    },
  },
  args: {
    name: 'eth',
    src: './images/eth_logo.png',
    size: Size.MD,
    showHalo: false,
  },
};

const Template = (args) => {
  return <AvatarToken {...args} />;
};

export const DefaultStory = Template.bind({});
DefaultStory.storyName = 'Default';

export const SizeStory = (args) => (
  <>
    <Box
      display={DISPLAY.FLEX}
      alignItems={AlignItems.baseline}
      gap={2}
      marginBottom={4}
    >
      <AvatarToken {...args} size={Size.XS} />
      <AvatarToken {...args} size={Size.SM} />
      <AvatarToken {...args} size={Size.MD} />
      <AvatarToken {...args} size={Size.LG} />
      <AvatarToken {...args} size={Size.XL} />
    </Box>
    <Box
      display={DISPLAY.FLEX}
      alignItems={AlignItems.flexEnd}
      gap={2}
      marginBottom={4}
    >
      <AvatarToken {...args} src="" size={Size.XS} />
      <AvatarToken {...args} src="" size={Size.SM} />
      <AvatarToken {...args} src="" size={Size.MD} />
      <AvatarToken {...args} src="" size={Size.LG} />
      <AvatarToken {...args} src="" size={Size.XL} />
    </Box>
    <Text marginBottom={4}>
      Sizes with{' '}
      <ButtonLink
        size={BUTTON_LINK_SIZES.INHERIT}
        href="/docs/components-componentlibrary-buttonlink--default-story"
      >
        AvatarNetwork
      </ButtonLink>{' '}
      and{' '}
      <ButtonLink
        size={BUTTON_LINK_SIZES.INHERIT}
        href="/docs/components-componentlibrary-badgewrapper--default-story"
      >
        BadgeWrapper
      </ButtonLink>{' '}
      components
    </Text>
    <Box
      display={DISPLAY.FLEX}
      alignItems={AlignItems.flexEnd}
      gap={2}
      marginBottom={4}
    >
      <BadgeWrapper
        badge={
          <AvatarNetwork
            src="./images/eth_logo.png"
            name="ETH"
            size={Size.XS}
            borderColor={BackgroundColor.backgroundDefault}
            borderWidth={2}
          />
        }
      >
        <AvatarToken {...args} name="ETH" size={Size.XS} />
      </BadgeWrapper>
      <BadgeWrapper
        badge={
          <AvatarNetwork
            src="./images/eth_logo.png"
            name="ETH"
            size={Size.XS}
            borderColor={BackgroundColor.backgroundDefault}
            borderWidth={2}
          />
        }
      >
        <AvatarToken {...args} name="ETH" size={Size.SM} />
      </BadgeWrapper>
      <BadgeWrapper
        badge={
          <AvatarNetwork
            src="./images/eth_logo.png"
            name="ETH"
            size={Size.XS}
            borderColor={BackgroundColor.backgroundDefault}
            borderWidth={2}
          />
        }
      >
        <AvatarToken {...args} name="ETH" size={Size.MD} />
      </BadgeWrapper>
      <BadgeWrapper
        badge={
          <AvatarNetwork
            src="./images/eth_logo.png"
            name="ETH"
            size={Size.XS}
            borderColor={BackgroundColor.backgroundDefault}
            borderWidth={2}
          />
        }
      >
        <AvatarToken {...args} name="ETH" size={Size.LG} />
      </BadgeWrapper>
      <BadgeWrapper
        badge={
          <AvatarNetwork
            src="./images/eth_logo.png"
            name="ETH"
            size={Size.SM}
            borderColor={BackgroundColor.backgroundDefault}
            borderWidth={2}
          />
        }
      >
        <AvatarToken {...args} name="ETH" size={Size.XL} />
      </BadgeWrapper>
    </Box>
    <Box display={DISPLAY.FLEX} alignItems={AlignItems.flexEnd} gap={2}>
      <BadgeWrapper
        badge={
          <AvatarNetwork
            name="ETH"
            size={Size.XS}
            borderColor={BackgroundColor.backgroundDefault}
            borderWidth={2}
          />
        }
      >
        <AvatarToken
          {...args}
          src=""
          name="ETH"
          size={Size.XS}
          borderColor={BorderColor.borderDefault}
          borderSize={2}
        />
      </BadgeWrapper>
      <BadgeWrapper
        badge={
          <AvatarNetwork
            name="ETH"
            size={Size.XS}
            borderColor={BackgroundColor.backgroundDefault}
            borderWidth={2}
          />
        }
      >
        <AvatarToken
          {...args}
          name="ETH"
          src=""
          size={Size.SM}
          borderColor={BorderColor.borderDefault}
          borderSize={2}
        />
      </BadgeWrapper>
      <BadgeWrapper
        badge={
          <AvatarNetwork
            name="ETH"
            size={Size.XS}
            borderColor={BackgroundColor.backgroundDefault}
            borderWidth={2}
          />
        }
      >
        <AvatarToken
          {...args}
          name="ETH"
          src=""
          size={Size.MD}
          borderColor={BorderColor.borderDefault}
          borderSize={2}
        />
      </BadgeWrapper>
      <BadgeWrapper
        badge={
          <AvatarNetwork
            name="ETH"
            size={Size.XS}
            borderColor={BackgroundColor.backgroundDefault}
            borderWidth={2}
          />
        }
      >
        <AvatarToken
          {...args}
          name="ETH"
          src=""
          size={Size.LG}
          borderColor={BorderColor.borderDefault}
          borderSize={2}
        />
      </BadgeWrapper>
      <BadgeWrapper
        badge={
          <AvatarNetwork
            name="ETH"
            size={Size.SM}
            borderColor={BackgroundColor.backgroundDefault}
            borderWidth={2}
          />
        }
      >
        <AvatarToken
          {...args}
          name="ETH"
          src=""
          size={Size.XL}
          borderColor={BorderColor.borderDefault}
          borderSize={2}
        />
      </BadgeWrapper>
    </Box>
  </>
);
SizeStory.storyName = 'Size';

export const Name = Template.bind({});
Name.args = {
  src: '',
};

export const Src = (args) => (
  <Box display={DISPLAY.FLEX} gap={1}>
    <AvatarToken {...args} src="./images/eth_logo.png" />
    <AvatarToken {...args} src="./images/arbitrum.svg" />
    <AvatarToken {...args} src="./images/bnb.png" />
    <AvatarToken
      {...args}
      src="https://static.metaswap.codefi.network/api/v1/tokenIcons/1/0x6b175474e89094c44da98b954eedeac495271d0f.png"
    />
    <AvatarToken
      {...args}
      src="https://static.metaswap.codefi.network/api/v1/tokenIcons/1/0x0d8775f648430679a709e98d2b0cb6250d2887ef.png"
    />
    <AvatarToken
      {...args}
      src="https://static.metaswap.codefi.network/api/v1/tokenIcons/1/0x7d1afa7b718fb893db30a3abc0cfc608aacfebb0.png"
    />
    <AvatarToken
      {...args}
      src="https://i.seadn.io/gae/lSm8ChaI-3RqC9MTpi0j3KBXdfdPd57PN5UeQLY49JA3twy9wSt2dpaa22sSc6oyiXi2OEUR6GeFX8jwkZHEMADu6_Bd4EwTJ-rg?w=500&auto=format"
    />
  </Box>
);

export const ShowHalo = Template.bind({});
ShowHalo.args = {
  showHalo: true,
};

export const ColorBackgroundColorAndBorderColor = (args) => (
  <Box display={DISPLAY.FLEX} gap={1}>
    <AvatarToken
      {...args}
      backgroundColor={BackgroundColor.goerli}
      borderColor={BorderColor.goerli}
      name="G"
      color={TextColor.primaryInverse}
    />
    <AvatarToken
      {...args}
      backgroundColor={BackgroundColor.sepolia}
      borderColor={BorderColor.sepolia}
      name="S"
      color={TextColor.primaryInverse}
    />
  </Box>
);
ColorBackgroundColorAndBorderColor.args = {
  src: '',
};
