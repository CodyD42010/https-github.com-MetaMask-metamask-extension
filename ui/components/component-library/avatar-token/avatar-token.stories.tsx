import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import {
  DISPLAY,
  AlignItems,
  TextColor,
  BackgroundColor,
  BorderColor,
} from '../../../helpers/constants/design-system';

import Box from '../../ui/box/box';

import {
  AvatarNetwork,
  BUTTON_LINK_SIZES,
  BadgeWrapper,
  ButtonLink,
  Text,
} from '..';

import { AvatarBaseSize } from '../avatar-base/avatar-base.types';
import README from './README.mdx';

import { AvatarToken } from './avatar-token';

export default {
  title: 'Components/ComponentLibrary/AvatarToken',
  component: AvatarToken,
  parameters: {
    docs: {
      page: README,
    },
  },
  argTypes: {
    size: {
      control: 'select',
      options: Object.values(AvatarBaseSize),
    },
    color: {
      options: Object.values(TextColor),
      control: 'select',
    },
    backgroundColor: {
      options: Object.values(BackgroundColor),
      control: 'select',
    },
    borderColor: {
      options: Object.values(BorderColor),
      control: 'select',
    },
    name: {
      control: 'text',
    },
    src: {
      control: 'text',
    },
    showHalo: {
      control: 'boolean',
    },
  },
  args: {
    name: 'eth',
    src: './images/eth_logo.png',
    size: AvatarBaseSize.Md,
    showHalo: false,
  },
} as ComponentMeta<typeof AvatarToken>;

const Template: ComponentStory<typeof AvatarToken> = (args) => {
  return <AvatarToken {...args} />;
};

export const DefaultStory = Template.bind({});
DefaultStory.storyName = 'Default';

export const SizeStory: ComponentStory<typeof AvatarToken> = (args) => (
  <>
    <Box
      display={DISPLAY.FLEX}
      alignItems={AlignItems.baseline}
      gap={2}
      marginBottom={4}
    >
      <AvatarToken {...args} size={AvatarBaseSize.Xs} />
      <AvatarToken {...args} size={AvatarBaseSize.Sm} />
      <AvatarToken {...args} size={AvatarBaseSize.Md} />
      <AvatarToken {...args} size={AvatarBaseSize.Lg} />
      <AvatarToken {...args} size={AvatarBaseSize.Xl} />
    </Box>
    <Box
      display={DISPLAY.FLEX}
      alignItems={AlignItems.flexEnd}
      gap={2}
      marginBottom={4}
    >
      <AvatarToken {...args} src="" size={AvatarBaseSize.Xs} />
      <AvatarToken {...args} src="" size={AvatarBaseSize.Sm} />
      <AvatarToken {...args} src="" size={AvatarBaseSize.Md} />
      <AvatarToken {...args} src="" size={AvatarBaseSize.Lg} />
      <AvatarToken {...args} src="" size={AvatarBaseSize.Xl} />
    </Box>
    <Text marginBottom={4}>
      Sizes with{' '}
      <ButtonLink
        size={BUTTON_LINK_SIZES.INHERIT}
        href="/docs/components-componentlibrary-buttonlink--default-story"
      >
        AvatarNetwork
      </ButtonLink>{' '}
      and{' '}
      <ButtonLink
        size={BUTTON_LINK_SIZES.INHERIT}
        href="/docs/components-componentlibrary-badgewrapper--default-story"
      >
        BadgeWrapper
      </ButtonLink>{' '}
      components
    </Text>
    <Box
      display={DISPLAY.FLEX}
      alignItems={AlignItems.flexEnd}
      gap={2}
      marginBottom={4}
    >
      <BadgeWrapper
        badge={
          <AvatarNetwork
            src="./images/eth_logo.png"
            name="ETH"
            size={AvatarBaseSize.Xs}
            borderColor={BackgroundColor.backgroundDefault}
            borderWidth={2}
          />
        }
      >
        <AvatarToken {...args} name="ETH" size={AvatarBaseSize.Xs} />
      </BadgeWrapper>
      <BadgeWrapper
        badge={
          <AvatarNetwork
            src="./images/eth_logo.png"
            name="ETH"
            size={AvatarBaseSize.Xs}
            borderColor={BackgroundColor.backgroundDefault}
            borderWidth={2}
          />
        }
      >
        <AvatarToken {...args} name="ETH" size={AvatarBaseSize.Sm} />
      </BadgeWrapper>
      <BadgeWrapper
        badge={
          <AvatarNetwork
            src="./images/eth_logo.png"
            name="ETH"
            size={AvatarBaseSize.Xs}
            borderColor={BackgroundColor.backgroundDefault}
            borderWidth={2}
          />
        }
      >
        <AvatarToken {...args} name="ETH" size={AvatarBaseSize.Md} />
      </BadgeWrapper>
      <BadgeWrapper
        badge={
          <AvatarNetwork
            src="./images/eth_logo.png"
            name="ETH"
            size={AvatarBaseSize.Xs}
            borderColor={BackgroundColor.backgroundDefault}
            borderWidth={2}
          />
        }
      >
        <AvatarToken {...args} name="ETH" size={AvatarBaseSize.Lg} />
      </BadgeWrapper>
      <BadgeWrapper
        badge={
          <AvatarNetwork
            src="./images/eth_logo.png"
            name="ETH"
            size={AvatarBaseSize.Sm}
            borderColor={BackgroundColor.backgroundDefault}
            borderWidth={2}
          />
        }
      >
        <AvatarToken {...args} name="ETH" size={AvatarBaseSize.Xl} />
      </BadgeWrapper>
    </Box>
    <Box display={DISPLAY.FLEX} alignItems={AlignItems.flexEnd} gap={2}>
      <BadgeWrapper
        badge={
          <AvatarNetwork
            name="ETH"
            size={AvatarBaseSize.Xs}
            borderColor={BackgroundColor.backgroundDefault}
            borderWidth={2}
          />
        }
      >
        <AvatarToken
          {...args}
          src=""
          name="ETH"
          size={AvatarBaseSize.Xs}
          borderColor={BorderColor.borderDefault}
          borderSize={2}
        />
      </BadgeWrapper>
      <BadgeWrapper
        badge={
          <AvatarNetwork
            name="ETH"
            size={AvatarBaseSize.Xs}
            borderColor={BackgroundColor.backgroundDefault}
            borderWidth={2}
          />
        }
      >
        <AvatarToken
          {...args}
          name="ETH"
          src=""
          size={AvatarBaseSize.Sm}
          borderColor={BorderColor.borderDefault}
          borderSize={2}
        />
      </BadgeWrapper>
      <BadgeWrapper
        badge={
          <AvatarNetwork
            name="ETH"
            size={AvatarBaseSize.Xs}
            borderColor={BackgroundColor.backgroundDefault}
            borderWidth={2}
          />
        }
      >
        <AvatarToken
          {...args}
          name="ETH"
          src=""
          size={AvatarBaseSize.Md}
          borderColor={BorderColor.borderDefault}
          borderSize={2}
        />
      </BadgeWrapper>
      <BadgeWrapper
        badge={
          <AvatarNetwork
            name="ETH"
            size={AvatarBaseSize.Xs}
            borderColor={BackgroundColor.backgroundDefault}
            borderWidth={2}
          />
        }
      >
        <AvatarToken
          {...args}
          name="ETH"
          src=""
          size={AvatarBaseSize.Lg}
          borderColor={BorderColor.borderDefault}
          borderSize={2}
        />
      </BadgeWrapper>
      <BadgeWrapper
        badge={
          <AvatarNetwork
            name="ETH"
            size={AvatarBaseSize.Sm}
            borderColor={BackgroundColor.backgroundDefault}
            borderWidth={2}
          />
        }
      >
        <AvatarToken
          {...args}
          name="ETH"
          src=""
          size={AvatarBaseSize.Xl}
          borderColor={BorderColor.borderDefault}
          borderSize={2}
        />
      </BadgeWrapper>
    </Box>
  </>
);
SizeStory.storyName = 'Size';

export const Name = Template.bind({});
Name.args = {
  src: '',
};

export const Src: ComponentStory<typeof AvatarToken> = (args) => (
  <Box display={DISPLAY.FLEX} gap={1}>
    <AvatarToken {...args} src="./images/eth_logo.png" />
    <AvatarToken {...args} src="./images/arbitrum.svg" />
    <AvatarToken {...args} src="./images/bnb.png" />
    <AvatarToken
      {...args}
      src="https://static.metaswap.codefi.network/api/v1/tokenIcons/1/0x6b175474e89094c44da98b954eedeac495271d0f.png"
    />
    <AvatarToken
      {...args}
      src="https://static.metaswap.codefi.network/api/v1/tokenIcons/1/0x0d8775f648430679a709e98d2b0cb6250d2887ef.png"
    />
    <AvatarToken
      {...args}
      src="https://static.metaswap.codefi.network/api/v1/tokenIcons/1/0x7d1afa7b718fb893db30a3abc0cfc608aacfebb0.png"
    />
    <AvatarToken
      {...args}
      src="https://i.seadn.io/gae/lSm8ChaI-3RqC9MTpi0j3KBXdfdPd57PN5UeQLY49JA3twy9wSt2dpaa22sSc6oyiXi2OEUR6GeFX8jwkZHEMADu6_Bd4EwTJ-rg?w=500&auto=format"
    />
  </Box>
);

export const ShowHalo = Template.bind({});
ShowHalo.args = {
  showHalo: true,
};

export const ColorBackgroundColorAndBorderColor: ComponentStory<
  typeof AvatarToken
> = (args) => (
  <Box display={DISPLAY.FLEX} gap={1}>
    <AvatarToken
      {...args}
      backgroundColor={BackgroundColor.goerli}
      borderColor={BorderColor.goerli}
      name="G"
      color={TextColor.primaryInverse}
    />
    <AvatarToken
      {...args}
      backgroundColor={BackgroundColor.sepolia}
      borderColor={BorderColor.sepolia}
      name="S"
      color={TextColor.primaryInverse}
    />
  </Box>
);
ColorBackgroundColorAndBorderColor.args = {
  src: '',
};
