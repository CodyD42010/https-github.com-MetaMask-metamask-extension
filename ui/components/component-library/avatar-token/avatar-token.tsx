import React, { useState, useEffect, Ref } from 'react';
import classnames from 'classnames';
import { AvatarBase, AvatarBaseSize } from '../avatar-base';
import {
  DISPLAY,
  AlignItems,
  JustifyContent,
  TextColor,
  BorderColor,
  BackgroundColor,
} from '../../../helpers/constants/design-system';
import { AvatarTokenProps } from './avatar-token.types';

export const AvatarToken = React.forwardRef(
  (
    {
      size = AvatarBaseSize.Md,
      name,
      src,
      showHalo,
      color = TextColor.textDefault,
      backgroundColor = BackgroundColor.backgroundAlternative,
      borderColor = BorderColor.transparent,
      className = '',
      ...props
    }: AvatarTokenProps,
    ref: Ref<HTMLElement>,
  ) => {
    const [showFallback, setShowFallback] = useState(false);

    useEffect(() => {
      setShowFallback(!src);
    }, [src]);

    const handleOnError = () => {
      setShowFallback(true);
    };

    const fallbackString = name?.[0] ?? '?';

    return (
      <AvatarBase
        ref={ref}
        size={size}
        display={DISPLAY.FLEX}
        alignItems={AlignItems.center}
        justifyContent={JustifyContent.center}
        className={classnames(
          'mm-avatar-token',
          showHalo ? 'mm-avatar-token--with-halo' : '',
          className,
        )}
        {...{ backgroundColor, borderColor, color, ...props }}
      >
        {showFallback ? (
          fallbackString
        ) : (
          <>
            {showHalo && (
              <img
                src={src}
                className={
                  showHalo ? 'mm-avatar-token__token-image--blurred' : ''
                }
                aria-hidden="true"
              />
            )}
            <img
              className={
                showHalo
                  ? 'mm-avatar-token__token-image--size-reduced'
                  : 'mm-avatar-token__token-image'
              }
              onError={handleOnError}
              src={src}
              alt={`${name} logo` || 'token logo'}
            />
          </>
        )}
      </AvatarBase>
    );
  },
);

AvatarToken.displayName = 'AvatarToken';
