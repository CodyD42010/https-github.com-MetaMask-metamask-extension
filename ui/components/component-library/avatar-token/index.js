import { AvatarToken } from "./avatar-token";