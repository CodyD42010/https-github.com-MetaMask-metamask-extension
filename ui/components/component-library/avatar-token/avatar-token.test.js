/* eslint-disable jest/require-top-level-describe */
import { render, screen } from '@testing-library/react';
import React from 'react';

import {
  BackgroundColor,
  BorderColor,
  TextColor,
} from '../../../helpers/constants/design-system';

import { AvatarToken } from './avatar-token';

describe('AvatarToken', () => {
  const args = {
    name: 'ast',
    src: './AST.png',
    showHalo: false,
  };

  it('should render correctly', () => {
    const { getByTestId, container } = render(
      <AvatarToken {...args} data-testid="avatar-token" />,
    );
    expect(getByTestId('avatar-token')).toBeDefined();
    expect(container).toMatchSnapshot();
  });

  it('should render image Avatar', () => {
    render(<AvatarToken {...args} data-testid="avatar-token" />);
    const image = screen.getByRole('img');
    expect(image).toBeDefined();
    expect(image).toHaveAttribute('src', args.src);
  });

  it('should render the first letter of the name prop if no src is provided', () => {
    const { getByText } = render(
      <AvatarToken {...args} data-testid="avatar-token" src="" />,
    );
    expect(getByText('a')).toBeDefined();
  });

  it('should render halo effect if showHalo is true and image url is there', () => {
    render(<AvatarToken {...args} data-testid="avatar-token" showHalo />);
    const image = screen.getAllByRole('img', { hidden: true });
    expect(image[1]).toHaveClass('mm-avatar-token__token-image--size-reduced');
  });

  it('should render the first letter of the name prop when showHalo is true and no image url is provided', () => {
    const { getByText } = render(
      <AvatarToken {...args} src="" data-testid="avatar-token" showHalo />,
    );
    expect(getByText('a')).toBeDefined();
  });
  // className
  it('should render with custom className', () => {
    const { getByTestId } = render(
      <AvatarToken data-testid="avatar-token" className="test-class" />,
    );
    expect(getByTestId('avatar-token')).toHaveClass('test-class');
  });
  // color
  it('should render with different colors', () => {
    const { getByTestId } = render(
      <>
        <AvatarToken
          color={TextColor.successDefault}
          data-testid={TextColor.successDefault}
        />
        <AvatarToken
          color={TextColor.errorDefault}
          data-testid={TextColor.errorDefault}
        />
      </>,
    );
    expect(getByTestId(TextColor.successDefault)).toHaveClass(
      `box--color-${TextColor.successDefault}`,
    );
    expect(getByTestId(TextColor.errorDefault)).toHaveClass(
      `box--color-${TextColor.errorDefault}`,
    );
  });
  // background color
  it('should render with different background colors', () => {
    const { getByTestId } = render(
      <>
        <AvatarToken
          backgroundColor={BackgroundColor.successDefault}
          data-testid={BackgroundColor.successDefault}
        />
        <AvatarToken
          backgroundColor={BackgroundColor.errorDefault}
          data-testid={BackgroundColor.errorDefault}
        />
      </>,
    );
    expect(getByTestId(BackgroundColor.successDefault)).toHaveClass(
      `box--background-color-${BackgroundColor.successDefault}`,
    );
    expect(getByTestId(BackgroundColor.errorDefault)).toHaveClass(
      `box--background-color-${BackgroundColor.errorDefault}`,
    );
  });
  // border color
  it('should render with different border BorderColor', () => {
    const { getByTestId } = render(
      <>
        <AvatarToken
          borderColor={BorderColor.successDefault}
          data-testid={BorderColor.successDefault}
        />
        <AvatarToken
          borderColor={BorderColor.errorDefault}
          data-testid={BorderColor.errorDefault}
        />
      </>,
    );
    expect(getByTestId(BorderColor.successDefault)).toHaveClass(
      `box--border-color-${BorderColor.successDefault}`,
    );
    expect(getByTestId(BorderColor.errorDefault)).toHaveClass(
      `box--border-color-${BorderColor.errorDefault}`,
    );
  });
  it('should forward a ref to the root html element', () => {
    const ref = React.createRef();
    render(<AvatarToken ref={ref} />);
    expect(ref.current).not.toBeNull();
    expect(ref.current.nodeName).toBe('DIV');
  });
});
