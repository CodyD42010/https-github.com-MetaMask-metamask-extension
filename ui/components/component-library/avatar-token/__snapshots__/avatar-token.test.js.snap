// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarToken should render correctly 1`] = `
<div>
  <div
    class="box mm-text mm-avatar-base mm-avatar-base--size-md mm-avatar-token mm-text--body-sm mm-text--text-transform-uppercase box--display-flex box--flex-direction-row box--justify-content-center box--align-items-center box--color-text-default box--background-color-background-alternative box--rounded-full box--border-color-transparent box--border-style-solid box--border-width-1"
    data-testid="avatar-token"
  >
    <img
      alt="ast logo"
      class="mm-avatar-token__token-image"
      src="./AST.png"
    />
  </div>
</div>
`;
